test/common/parallel_tests.js

```diff
@@ -57,101 +77,168 @@ function parallelTests(tag, suite, section) {
+				'--',
+				'--grep',
+				'@slow|@unstable|@standalone',
+				'--invert'
+			);
 			break;
 	}
+	return mochaArguments;
+};
 
-	// Looking recursevely for javascript files not containing the word "common"
-	var filepaths = find.fileSync(/^((?!common)[\s\S])*.js$/, suiteFolder);
-	var initFilepaths = filepaths.splice(0, maxParallelism);
-
-	var parallelTestsRunning = {};
-
-	var spawnTest = function(test) {
-		var coverageArguments = [
-			'cover',
-			'--dir',
-			'test/.coverage-unit',
-			'--include-pid',
-			'--print',
-			'none',
-			'node_modules/.bin/_mocha',
-			test,
-		];
-		var istanbulArguments = coverageArguments.concat(mochaArguments);
-
-		var child = child_process.spawn(
-			'node_modules/.bin/istanbul',
-			istanbulArguments,
-			{
-				cwd: `${__dirname}/../..`,
-				detached: true,
-				stdio: 'inherit',
-			}
-		);
+const spawnParallelTest = (testFile, mochaArguments) => {
+	return new Promise((resolve, reject) => {
+		const child = executeIstanbul(testFile, mochaArguments);
 
 		console.info(
-			'Running the test:',
-			test,
-			'as a separate process - pid',
-			child.pid
+			`Running parallel the test: ${testFile} as a separate process - pid: ${
+				child.pid
+			}`
 		);
-		parallelTestsRunning[child.pid] = child;
 
-		var cleanupRunningTests = function() {
-			Object.keys(parallelTestsRunning).forEach(k => {
-				parallelTestsRunning[k].kill('SIGTERM');
-			});
+		child.on('close', code => {
+			if (code === 0) {
+				console.info(`Test finished successfully: ${testFile}`);
+				resolve(testFile);
+				return child.kill('SIGTERM');
+			}
+
+			console.error('Test failed:', testFile);
+			reject(code);
+			return process.exit(code);
+		});
+
+		child.on('error', err => {
+			console.error(err);
+			reject(err);
+			process.exit();
+		});
+	});
+};
+
+const runParallelTests = (suiteFolder, mochaArguments) => {
+	return new Promise((resolve, reject) => {
+		// Looking recursevely for javascript files not containing the word "common"
+		const allFiles = find.fileSync(/^((?!common)[\s\S])*.js$/, suiteFolder);
+		const allFilesLength = allFiles.length;
+
+		const completedFiles = {};
+
+		const next = () => {
+			const testFile = allFiles.splice(0, 1);
+			spawnParallelTest(testFile, mochaArguments)
+				.then(testFile => {
+					completedFiles[testFile] = 'done';
+					if (
+						allFiles.length === 0 &&
+						Object.keys(completedFiles).length === allFilesLength
+					) {
+						console.info('All parallel tests finished successfully.');
+						return resolve('All parallel tests finished successfully.');
+					}
+
+					if (allFiles.length > 0) {
+						return next();
+					}
+				})
+				.catch(err => {
+					console.error(`Parallel test failed: ${testFile}`);
+					return reject(err);
+				});
 		};
 
+		for (
+			let i = 0, limit = Math.min(allFilesLength, maxParallelism);
+			i < limit;
+			i += 1
+		) {
+			next();
+		}
+	});
+};
+
+const runStandaloneTests = (suiteFolder, mochaArguments) => {
```

Comment:
To me `runSequentialTests` makes more sense to explain the behavior. Point is we can run any test (specified by any tag) sequential .